### PR TITLE
feat: Make cosign copy faster

### PR DESCRIPTION
**Pull request by jonjohnsonjr**
#### Summary

I noticed that `cosign copy` was remarkably slow, so I bumped the go-containerregistry dependency to pick up some work I've been doing recently to simplify fixing that slowness.

I tested this by running `cosign copy` twice to copy `gcr.io/projectsigstore/cosign:v1.13.0` to Artifact Registry in us-central1. The second push is slower because it doesn't actually have to copy anything.

We're roughly 3x faster after this change.

|           | First Push | Second Push |
|-----------|------------|-------------|
| Before    | 69s        | 23s         |
| After     | **21s**    | **8s**      |
| Reduction | 70%        | 65%         |

There's more we can do, but this is an easy place to start.

#### Release Note

`cosign copy` was optimized to copy images faster.

### CODE REVIEW

File: cmd/cosign/cli/copy/copy.go [hunk 1]
```diff
@@ -109,15 +124,7 @@ func descriptorsEqual(a, b *v1.Descriptor) bool {
...
-func copyTagImage(tm tagMap, srcDigest name.Digest, dstRepo name.Repository, overwrite bool, opts ...remote.Option) error {
-	src, err := tm(srcDigest, ociremote.WithRemoteOptions(opts...))
-	if err != nil {
-		return err
-	}
-	return copyImage(src, dstRepo.Tag(src.Identifier()), overwrite, opts...)
-}
-
-func copyImage(src, dest name.Reference, overwrite bool, opts ...remote.Option) error {
+func copyImage(ctx context.Context, pusher *remote.Pusher, src, dest name.Reference, overwrite bool, opts ...remote.Option) error {
```
Comment (znewman01) [inline cmd/cosign/cli/copy/copy.go:127]: Nit: seems like this copies more than "images." Should it have a new name?

File: cmd/cosign/cli/copy/copy.go [hunk 2]
```diff
@@ -48,41 +50,49 @@ func CopyCmd(ctx context.Context, regOpts options.RegistryOptions, srcImg, dstIm
...
-		}
+			dst := dstRepoRef.Tag(src.Identifier())
+			g.Go(func() error {
+				return copyImage(ctx, pusher, src, dst, force, remoteOpts...)
+			})
 
-		// Copy SBOMs
-		if err := copyTagImage(ociremote.SBOMTag, srcDigest, dstRepoRef, force, remoteOpts...); err != nil {
-			return err
+			if i == 0 && sigOnly {
```
Comment (znewman01) [inline cmd/cosign/cli/copy/copy.go:86]: Nit: I don't love that this depends on the order in the `tagMap`.

My preference might be to have a small helper function like:

```go
func whichTagsToCopyFromArgs(sigOnly bool) []tagMap {
  if sigOnly {
    return []tagMap{ociremote.SignatureTag}
  }
  return []tagMap{ociremote.SignatureTag, ociremote.AttestationTag, ociremote.SBOMTag}
}
```

This is extremely forwards-compatible with https://github.com/sigstore/cosign/issues/2002 and IMO properly separates parsing args from "what does this command do?"

**Review by znewman01 — APPROVED** [review]
\m/

Left some nitpicks, feel free to leave as TODOs or ignore entirely!

File: cmd/cosign/cli/copy/copy.go (hunk 2 above)
Comment (jonjohnsonjr) [reply cmd/cosign/cli/copy/copy.go]: Not quite as simple as your helper (we avoid pushing the entity itself as well), but added something like that.

File: cmd/cosign/cli/copy/copy.go (hunk 1 above)
Comment (jonjohnsonjr) [reply cmd/cosign/cli/copy/copy.go]: I propose `copy` :)

**Review by znewman01 — APPROVED** [review]
LGTM thanks!

**Review by znewman01 — APPROVED** [review]

**Pull request merged**